### PR TITLE
Zosma migration

### PR DESCRIPTION
**Pull request by thumbnail**
* Publishing set of current connected controllers
 * Transforming a controller set + controller mapping into a map of something else. (vague, but used to transform the impl into a strategy)
 * Transformer which changes assignment (e.g. transforming a controller set + controller mapping into a new controller mapping).

todo
 - [ ] Naming of `ControllerTransformer`

```java
  @Value
  public static class Frame<C extends Controller> implements Controller.SetSupplier<C>, Controller.MapSupplier<Player> {
    @Delegate
    private final Controller.SetSupplier<C> controllerSetSupplier;

    public Map<Player, Set<Integer>> getControllerMapping() {
      HashMap<Player, Set<Integer>> map = new HashMap<>();
      map.put(new Player(), new HashSet<>(Arrays.asList(0, 2, 3)));
      return map;
    }
  }
  public static class Player {}

  public static void main(String[] args) {
    Flowable.fromPublisher(new ControllerSetPublisher<>(5, 500, XboxController::new))
        .map(Frame::new)
        .publish(
            new ControllerTransformer<>(
                XboxController::getDpad,
                dpads -> dpads.stream().findFirst().get()))
        .subscribe(System.out::println);

    while(true) {}
  }
```

**note** of the ~1300 added lines, just ~255 are Java.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,13 +1,209 @@
-cmake-build-debug/
```
Comment (romni) [inline .gitignore:1]: why were these removed?

File: .gitignore [hunk 2]
```diff
@@ -1,13 +1,209 @@
...
+.scala_dependencies
+.worksheet
+### Example user template template
+### Example user template
+
+# IntelliJ project files
+.idea
 *.iml
-CMakeCache.txt
-cmake_install.cmake
```
Comment (romni) [inline .gitignore:8]: ditto

File: README.md [hunk 3]
```diff
@@ -1,27 +1,50 @@
...
-## Documentation
+## Dependency
+
+#### Maven
+
+```
+<dependency>
+    <groupId>io.leonis</groupId>
+    <artifactId>frosty</artifactId>
+    <version>0.0.1</version>
```
Comment (romni) [inline README.md:19]: `frosty` hasn't been published yet and there's an open PR for this (#17 )

File: README.md [hunk 4]
```diff
@@ -1,27 +1,50 @@
...
+parametrized using interface instead of class types. Unfortunately Java's `Supplier` can only be
+implemented once due to type erasure; this is why most data types in `frosty` have a nested
+`Supplier` (or `SetSupplier`, `MappingSupplier`, e.a.) interface. This also allows for unnecessary
+coupling as an function that which uses a dpad can work with any controller with a dpad.
+
+## Documentation
 
-The following packages are required:
-- [GCC](https://gcc.gnu.org/install/binaries.html)
-- [SDL (Simple Directmedia Layer) 2](https://www.libsdl.org/download-2.0.php)
```
Comment (romni) [inline README.md:17]: why was this removed?

File: src/main/java/io/leonis/frosty/ControllerTransformer.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * Function which reduces the accumulated set of O to O.
+   */
+  private final java.util.function.Function<Set<O>, O> outputReducer;
+
+  /**
+   * {@inheritDoc}
+   *
+   * todo this needs to be rewritten to use rxJava instead of streams
```
Comment (romni) [inline src/main/java/io/leonis/frosty/ControllerTransformer.java:40]: i see flowable is used. do you mean to rewrite the `.map` methods?

File: src/main/java/io/leonis/frosty/controllers/component/trigger/RightTriggerSupplier.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+ */
+public interface RightTriggerSupplier {
+
+  /**
+   * The amount the right trigger is pressed.
+   *
+   * @return A float between 0-1.
+   */
+  float getRightTrigger();
+}
```
Comment (romni) [inline src/main/java/io/leonis/frosty/controllers/component/trigger/RightTriggerSupplier.java:18]: missing newline

File: src/main/java/io/leonis/frosty/ControllerTransformer.java (hunk 5 above)
Comment (thumbnail) [reply src/main/java/io/leonis/frosty/ControllerTransformer.java:40]: Yes, that way backpressure works better i guess and we can possibly compile down to a lower version of java (without the streams lib)

File: .gitignore (hunk 1 above)
Comment (thumbnail) [reply .gitignore:1]: These files are no longer relevant to the project

File: .gitignore (hunk 2 above)
Comment (thumbnail) [reply .gitignore:8]: Ditto

File: README.md (hunk 4 above)
Comment (thumbnail) [reply README.md:17]: Jamepad supports all platforms we want and building the sdl impl posed problems on at Mac os and possibly windows as well. Since jamepad is pretty complete I propose just to use that and not deal with jni.

**Comment by thumbnail** [comment]
Implemented in https://github.com/delta-leonis/zosma/pull/41
